### PR TITLE
Create new DRYRUN_TEST_NUMBER

### PR DESCRIPTION
**Pull request by sastels**
# Summary | Résumé

Create a new DRYRUN_TEST_NUMBER that uses `dryrun=True` and switch the INTERNAL_TEST_NUMBER to just be internal (ie no boto call) as it used to be. This will allow us to load test the internals of Notify without having throttling errors (when we use the dryrun option).

## Related Issues | Cartes liées

* https://app.zenhub.com/workspaces/notify-planning-core-6411dfb7c95fb80014e0cab0/issues/gh/cds-snc/notification-planning-core/447

# Test instructions | Instructions pour tester la modification

- Send to INTERNAL_TEST_NUMBER = "+16135550123"
- should see in the logs something similar to:
```
[2024-11-04 20:24:18,291: INFO/ForkPoolWorker-5] notification d14da0e2-8c73-4cf3-a7a3-aa11141a6c86 is sending to INTERNAL_TEST_NUMBER, no boto call to AWS.
```
  - should not see a record in the AWS Pinpoint logs `PinpointDirectPublishToPhoneNumber`
- Send to DRYRUN_TEST_NUMBER = "+16135550124"
- should see in the logs something similar to:
```
[2024-11-04 20:22:13,332: INFO/ForkPoolWorker-5] SMS with message id 9792b3e8-2137-4a0d-8420-07bc6c325582 is sending to DRYRUN_TEST_NUMBER. Boto call made to AWS, but not send on.
```
- should not see a record in the AWS Pinpoint logs `PinpointDirectPublishToPhoneNumber` 

# Release Instructions | Instructions pour le déploiement

None.

# Reviewer checklist | Liste de vérification du réviseur

- [ ] This PR does not break existing functionality.
- [ ] This PR does not violate GCNotify's privacy policies.
- [ ] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
- [ ] This PR does not significantly alter performance.
- [ ] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: app/delivery/send_to_providers.py [hunk 1]
```diff
@@ -67,6 +67,10 @@ def send_sms_to_provider(notification):
         inactive_service_failure(notification=notification)
         return
 
+    formatted_recipient = validate_and_format_phone_number(notification.to, international=notification.international)
+    sending_to_internal_test_number = formatted_recipient == current_app.config["INTERNAL_TEST_NUMBER"]
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/delivery/send_to_providers.py:71]: it is

File: app/delivery/send_to_providers.py [hunk 2]
```diff
@@ -67,6 +67,10 @@ def send_sms_to_provider(notification):
...
+    sending_to_dryrun_number = formatted_recipient == current_app.config["DRYRUN_TEST_NUMBER"]
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/delivery/send_to_providers.py:72]: it is

File: app/clients/sms/aws_pinpoint.py [hunk 3]
```diff
@@ -52,7 +52,7 @@ def send_sms(self, to, content, reference, multi=True, sender=None, template_id=
                         ConfigurationSetName=self.current_app.config["AWS_PINPOINT_CONFIGURATION_SET_NAME"],
                     )
                 else:
-                    dryrun = destinationNumber == self.current_app.config["INTERNAL_TEST_NUMBER"]
+                    dryrun = destinationNumber == self.current_app.config["DRYRUN_TEST_NUMBER"]
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/clients/sms/aws_pinpoint.py:55]: hrmmmm it's defined up on line 42 - maybe we don't give that to the ai?

File: app/delivery/send_to_providers.py [hunk 4]
```diff
@@ -67,6 +67,10 @@ def send_sms_to_provider(notification):
         inactive_service_failure(notification=notification)
         return
 
+    formatted_recipient = validate_and_format_phone_number(notification.to, international=notification.international)
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/delivery/send_to_providers.py:70]: it's all good

File: app/delivery/send_to_providers.py [hunk 5]
```diff
@@ -125,10 +130,7 @@ def send_sms_to_provider(notification):
                 if reference == "opted_out":
                     update_notification_to_opted_out(notification, provider)
                 else:
-                    if (
-                        validate_and_format_phone_number(notification.to, international=notification.international)
-                        == current_app.config["INTERNAL_TEST_NUMBER"]
-                    ):
+                    if sending_to_dryrun_number:
                         send_sms_response(provider.get_name(), notification.to, reference)
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/delivery/send_to_providers.py:134]: up on line 72

File: tests/app/clients/test_aws_pinpoint.py [hunk 6]
```diff
@@ -158,20 +158,20 @@ def test_send_sms_sends_international_without_pool_id(notify_api, mocker, sample
...
 @pytest.mark.serial
 @pytest.mark.parametrize("template_id", [None, "uuid"])
-def test_send_sms_uses_dryrun_for_tests(notify_api, mocker, sample_template, template_id):
+def test_send_sms_uses_dryrun(notify_api, mocker, sample_template, template_id):
     boto_mock = mocker.patch.object(aws_pinpoint_client, "_client", create=True)
     mocker.patch.object(aws_pinpoint_client, "statsd_client", create=True)
     content = "foo"
     reference = "ref"
-    to = "+16135550123"
+    to = "+16135550111"
```
[the remark this answers is not in the document]
Comment (sastels) [reply tests/app/clients/test_aws_pinpoint.py:166]: no, we're explicitly setting the `DRYRUN_TEST_NUMBER` to this in a `with()`

File: tests/app/delivery/test_send_to_providers.py [hunk 7]
```diff
@@ -410,6 +410,24 @@ def test_should_not_send_sms_message_when_message_is_empty_or_whitespace(sample_
...
+            status="created",
+            reply_to_text=sample_service.get_default_sms_sender(),
+        )
+    )
+    mocker.patch("app.delivery.send_to_providers.send_sms_response", return_value="reference")
+    send_mock = mocker.patch("app.aws_sns_client.send_sms")
+    send_to_providers.send_sms_to_provider(notification)
+
+    send_mock.assert_not_called()
+    assert Notification.query.get(notification.id).status == "sent"
```
[the remark this answers is not in the document]
Comment (sastels) [reply tests/app/delivery/test_send_to_providers.py:428]: nope, we are pretending to send to this number, that's the whole point

File: app/delivery/send_to_providers.py [hunk 8]
```diff
@@ -93,7 +97,8 @@ def send_sms_to_provider(notification):
             empty_message_failure(notification=notification)
             return
 
-        if service.research_mode or notification.key_type == KEY_TYPE_TEST:
+        if service.research_mode or notification.key_type == KEY_TYPE_TEST or sending_to_internal_test_number:
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/delivery/send_to_providers.py:100]: defined up on line 71

**Review by ben851 — APPROVED** [review]
It looks like a thing!

**Pull request merged**